src/core/constraint.jl

```diff
@@ -32,6 +32,27 @@ function constraint_resistor_mass_flow(gm::AbstractGasModel, n::Int, k, f_min, f
     JuMP.set_upper_bound(f, ub)
 end
 
+#################################################################################################
+# Constraints associated with prvs
+#################################################################################################
+
+"Constraint: Constraints which define pressure drop across a prv"
+function constraint_prv_pressure(gm::AbstractGasModel, n::Int, k, i, j, pd)
+    pi = var(gm, n, :psqr, i)
+    pj = var(gm, n, :psqr, j)
+    _add_constraint!(gm, n, :pressure_drop, k, JuMP.@constraint(gm.model, pd == pi - pj))

```

Comment:
do prvs have a constant pressure drop?

Reply:
Yes, this is how these special resistors are defined in GasLib.

However, according to [this](https://en.wikipedia.org/wiki/Pressure_regulator), a pressure reducing valve only affects the pressure at the outlet node (`j`) and is insensitive to the pressure at the inlet node (`i`). Given that, maybe "pressure reducing valve" isn't the best nomenclature. Again, [GasLib documentation](https://www.mdpi.com/2306-5729/2/4/40/pdf) calls these components resistors (search for "pressureLoss"), which conflicts with how GasModels currently defines resistors. Maybe it would be better to rename "prv" to some special kind of resistor? I'm not quite sure.

Reply:
Yes, that was the source of my confusion. When I looked at the constraint, it didn't seem like prv type constraint.  I would recommend calling it some variation of a resistor... does the gaslib documentation or papers refer to it in any special way?  The other thing we could do is "type" the resistor like we do with compressors

Reply:
The GasLib paper states, "Two types of resistors appear in the GasLib instances. One results in a constant pressure loss (in the direction of the flow), which is specified by the element `pressureLoss`. The other type induces a pressure loss that continuously depends on the flow rate through the resistor as specified by the tags dragFactor and diameter."

Maybe we could refer to them as "constant" and "variable" resistors? Then I could just rename "prv" with "constant_resistor" everywhere. Or maybe "loss_resistor" or "drop_resistor?" I'm open to anything, really.

Reply:
given that the name resistor has been agreed to with ANL, lets keep that the way it is.  Lets rename prv to loss_resistor

Reply:
I'm fine with the name change, but thinking about this more, there's another issue: everything is in formulated in terms of squared pressure variables. However, loss resistors give a constant value for a pressure difference. So, what I currently have for a constraint right now, `loss^2 == p_i^2 - p_j^2`, is incorrect. Really, we want something that represents `loss^2 = (p_i - p_j)^2 = p_i^2 + p_j^2 - 2 * p_i * p_j`. How should we handle the bilinear term?

Reply:
Oh, very interesting.  So the NLP and MINLP should use the constraint directly since they allow nonlinear constraints. You can either keep everything in the square space (so have 2 * \sqr(p_i) * \sqrt(p_j)), or introduced a pressure variable and have a constraint like p^2 == psqr.  I am not sure which will be better behaved.  For the MISOCP, LP, and MIP, we'll need to relax this constraint.  You could relax it by dropping it completely (which is what we are currently doing for the compressor energy constraint-- though @kaarthiksundar and I are planning to revisit this later this year) and we'll come back to it.  Indeed, since we don't yet have p variables in the LP or MIP formulation, it should just be a blank constraint for those.  

The easiest relaxation for the MISOCP might be to have psqr >= p^2 and then do McCormick on p_i*p_j. (InfrastructureModels already as these relaxations implemented).  For now, since we don't want/need p variables in the non square space, I'd say we just create them locally in the constraint definition as auxiliary variables.... and if we start to run into more and more constraints where we need p in both the linear and square space, we'll come up with an architecture which has the smarts to know when to create the linear pressure variables and when not too...

Reply:
Another thing to think through.  Is directionality of flow a consideration here, i.e. does flow need to go in the same direction as the pressure drop? If the resistors are already one-way flow only, then we don't have to do much.  But if they are allowed to have two way flow, we'll need to enforce that.  Easy to do with the MINLP and NLP since we have direction variables.  With NLP, it is just f*(p_i-p_j) >= 0.  LP and MIP don't have p variables, so we are good.

Reply:
Want to take a look at the current formulations? The constraints are `constraint_loss_resistor_pressure`. A couple things: (i) in the NLP version, I use `JuMP.@constraint(gm.model, f * (p_i - p_j) == f * pd)`, which could behave unexpectedly when `f` is zero. Not sure how to get around this while also needing to direct the pressure difference; (ii) constraints on the pressure variables at `i` and `j` are both done within a constraint for an edge and thus may be redundantly added if two loss resistors are added in series (e.g., `JuMP.@constraint(gm.model, p_i^2 == p_i_sqr)` and `JuMP.@constraint(gm.model, p_j^2 == p_j_sqr)` are added in the same constraint for one loss resistor). I can break these pressure-squaring constraints out into new constraints per junction, if you'd like.

Reply:
I am not entirely sure what you are trying to model with ``f * (p_i - p_j) == f * pd)''  Are you trying to make sure that direction of flow is the same as the pressure drop?  In which case, you could do ``f*(pi-pj) >= 0'' and ''pd == abs(pi - pj)''.  This will make sure the sign of f is the same as the sign of pi-pj and make sure that pressure difference matches the sign of pi - pj.  If flow can only go one way, all you need is ``pd == pi -pj'' since somewhere else you will have ``f >=0''

Reply:
Ah, I didn't know `abs` could be used in the constraint like that. Yes, I was trying to model pressure drop and flow directions were the same. (The components appear to be bidirectional.) This has been updated to use your suggestions.

Reply:
Looks good.  I'll approve once all the checks pass.

Reply:
Actually, sorry, `abs(p_i - p_j)` is not valid. However, `(p_i - p_j)^2 == pd^2` is, so we will go with that for the NLP formulation.

Reply:
Really? I thought p_i, p_j, and pd were all in the space of linear pressure?  We use abs in the Weymouth formulation....

Reply:
Invalid meaning you can't squeeze in a difference of two variables in an `abs` operator.

Reply:
Ah, AutoDiff must not be able to handle it.  We could introduce an auxiliary variable, but your way also works,